File: code/go/03-agent-loop/llm_provider.go
// LLM Provider abstraction — Go port.
//
// Implements a uniform interface over multiple LLM providers so agent code
// never depends on a specific SDK or wire format.
//
// Supported providers:
//
//	OpenAIProvider, AnthropicProvider, OllamaProvider, FallbackProvider
//
// Usage:
//
//	provider := NewOpenAIProvider(os.Getenv("OPENAI_API_KEY"), "gpt-4o")
//	resp, err := provider.Chat(context.Background(), []Message{
//	    {Role: "user", Content: "Hello"},
//	}, ChatOptions{})
//
// See: docs/05-the-tool-ecosystem/01-model-providers.md
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"time"
)

// ---------------------------------------------------------------------------
// Shared data types
// ---------------------------------------------------------------------------

// Message is a single chat message in the OpenAI format.
type Message struct {
	Role       string     `json:"role"`
	Content    string     `json:"content"`
	ToolCallID string     `json:"tool_call_id,omitempty"`
	ToolCalls  []ToolCall `json:"tool_calls,omitempty"`
}

// ToolCall is a function-calling entry from an assistant message.
type ToolCall struct {
	ID       string           `json:"id"`
	Type     string           `json:"type"`
	Function ToolCallFunction `json:"function"`
}

// ToolCallFunction holds the name and JSON-encoded arguments.
type ToolCallFunction struct {
	Name      string `json:"name"`
	Arguments string `json:"arguments"`
}

// ToolDefinition is the OpenAI-format tool definition.
type ToolDefinition struct {
	Type     string             `json:"type"` // "function"
	Function ToolFunctionSchema `json:"function"`
}

// ToolFunctionSchema describes a function tool.
type ToolFunctionSchema struct {
	Name        string         `json:"name"`
	Description string         `json:"description,omitempty"`
	Parameters  map[string]any `json:"parameters,omitempty"`
}

// TokenUsage holds token consumption for a call.
type TokenUsage struct {
	PromptTokens     int `json:"prompt_tokens"`
	CompletionTokens int `json:"completion_tokens"`
	TotalTokens      int `json:"total_tokens"`
}

// LLMResponse is the normalised response returned by every provider.
type LLMResponse struct {
	Content      string     // nil-equivalent: empty string
	ToolCalls    []ToolCall // nil when no tool calls
	TokenUsage   TokenUsage
	Model        string
	FinishReason string
	LatencyMs    int
}

// ChatOptions holds optional parameters for a Chat call.
type ChatOptions struct {
	Tools          []ToolDefinition
	Temperature    float64
	MaxTokens      int
	ResponseFormat map[string]any
}

func defaultChatOptions(o ChatOptions) ChatOptions {
	if o.Temperature == 0 {
		o.Temperature = 0.7
	}
	if o.MaxTokens == 0 {
		o.MaxTokens = 4096
	}
	return o
}

// EstimateTokens returns a rough token estimate (~4 chars per token).
func EstimateTokens(text string) int {
	n := len(text) / 4
	if n < 1 {
		return 1
	}
	return n
}

// ---------------------------------------------------------------------------
// LLMProvider interface
// ---------------------------------------------------------------------------

// LLMProvider is the common interface for all LLM providers.
type LLMProvider interface {
	Chat(ctx context.Context, messages []Message, opts ChatOptions) (LLMResponse, error)
	SupportsFunctionCalling() bool
	SupportsStructuredOutput() bool
	GetContextWindow() int
	GetModelName() string
	CountTokens(text string) int
}

// ---------------------------------------------------------------------------
// OpenAI provider  (also used as the engine for Ollama / Together)
// ---------------------------------------------------------------------------

// OpenAIProvider talks to the OpenAI chat completions endpoint
// (or any OpenAI-compatible endpoint via BaseURL).
type OpenAIProvider struct {
	APIKey  string
	Model   string
	BaseURL string
	client  *http.Client
}

var openaiContextWindows = map[string]int{
	"gpt-4o":        128_000,
	"gpt-4o-mini":   128_000,
	"gpt-3.5-turbo": 16_385,
}

// NewOpenAIProvider creates an OpenAIProvider for the given model.
func NewOpenAIProvider(apiKey, model string) *OpenAIProvider {
	return &OpenAIProvider{
		APIKey: apiKey, Model: model,
		BaseURL: "https://api.openai.com/v1",
		client:  &http.Client{Timeout: 120 * time.Second},
	}
}

func (p *OpenAIProvider) Chat(ctx context.Context, messages []Message, opts ChatOptions) (LLMResponse, error) {
	opts = defaultChatOptions(opts)

	type oaiReq struct {
		Model          string           `json:"model"`
		Messages       []Message        `json:"messages"`
		Temperature    float64          `json:"temperature"`
		MaxTokens      int              `json:"max_tokens"`
		Tools          []ToolDefinition `json:"tools,omitempty"`
		ResponseFormat map[string]any   `json:"response_format,omitempty"`
	}
	type oaiChoice struct {
		Message      Message `json:"message"`
		FinishReason string  `json:"finish_reason"`
	}
	type oaiUsage struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
		TotalTokens      int `json:"total_tokens"`
	}
	type oaiResp struct {
		Choices []oaiChoice `json:"choices"`
		Usage   oaiUsage    `json:"usage"`
		Model   string      `json:"model"`
	}

	payload := oaiReq{
		Model:       p.Model,
		Messages:    messages,
		Temperature: opts.Temperature,
		MaxTokens:   opts.MaxTokens,
	}
	if len(opts.Tools) > 0 {
		payload.Tools = opts.Tools
	}
	if opts.ResponseFormat != nil {
		payload.ResponseFormat = opts.ResponseFormat
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return LLMResponse{}, fmt.Errorf("marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", p.BaseURL+"/chat/completions", bytes.NewReader(body))
	if err != nil {
		return LLMResponse{}, err
	}
	req.Header.Set("Content-Type", "application/json")
	if p.APIKey != "" && p.APIKey != "ollama" {
		req.Header.Set("Authorization", "Bearer "+p.APIKey)
	}

	t0 := time.Now()
	httpResp, err := p.client.Do(req)
	latencyMs := int(time.Since(t0).Milliseconds())
	if err != nil {
		return LLMResponse{}, fmt.Errorf("http request: %w", err)
	}
	defer httpResp.Body.Close()

	respBody, err := io.ReadAll(httpResp.Body)
	if err != nil {
		return LLMResponse{}, fmt.Errorf("read response: %w", err)
	}
	if httpResp.StatusCode >= 400 {
		return LLMResponse{}, fmt.Errorf("API error %d: %s", httpResp.StatusCode, respBody)
	}

	var parsed oaiResp
	if err := json.Unmarshal(respBody, &parsed); err != nil {
		return LLMResponse{}, fmt.Errorf("parse response: %w", err)
	}
	if len(parsed.Choices) == 0 {
		return LLMResponse{}, fmt.Errorf("empty choices in response")
	}

	msg := parsed.Choices[0].Message
	return LLMResponse{
		Content:      msg.Content,
		ToolCalls:    msg.ToolCalls,
		TokenUsage:   TokenUsage{PromptTokens: parsed.Usage.PromptTokens, CompletionTokens: parsed.Usage.CompletionTokens, TotalTokens: parsed.Usage.TotalTokens},
		Model:        p.Model,
		FinishReason: parsed.Choices[0].FinishReason,
		LatencyMs:    latencyMs,
	}, nil
}

func (p *OpenAIProvider) SupportsFunctionCalling() bool  { return true }
func (p *OpenAIProvider) SupportsStructuredOutput() bool { return true }
func (p *OpenAIProvider) GetContextWindow() int {
	if w, ok := openaiContextWindows[p.Model]; ok {
		return w
	}
	return 128_000
}
func (p *OpenAIProvider) GetModelName() string        { return p.Model }
func (p *OpenAIProvider) CountTokens(text string) int { return EstimateTokens(text) }

// ---------------------------------------------------------------------------
// Anthropic provider
// ---------------------------------------------------------------------------

// AnthropicProvider talks to the Anthropic Messages API.
type AnthropicProvider struct {
	APIKey string
	Model  string
	client *http.Client
}

// NewAnthropicProvider creates an AnthropicProvider.
func NewAnthropicProvider(apiKey, model string) *AnthropicProvider {
	return &AnthropicProvider{
		APIKey: apiKey, Model: model,
		client: &http.Client{Timeout: 120 * time.Second},
	}
}

// toAnthropicMessages converts OpenAI-format messages to Anthropic format.
func toAnthropicMessages(messages []Message) (system string, converted []map[string]any) {
	for _, m := range messages {
		switch m.Role {
		case "system":
			system = m.Content
		case "tool":
			converted = append(converted, map[string]any{
				"role": "user",
				"content": []map[string]any{{
					"type":        "tool_result",
					"tool_use_id": m.ToolCallID,
					"content":     m.Content,
				}},
			})
		default:
			converted = append(converted, map[string]any{
				"role":    m.Role,
				"content": m.Content,
			})
		}
	}
	return
}

// toAnthropicTools converts OpenAI tool definitions to Anthropic input_schema format.
func toAnthropicTools(tools []ToolDefinition) []map[string]any {
	out := make([]map[string]any, len(tools))
	for i, t := range tools {
		params := t.Function.Parameters
		if params == nil {
			params = map[string]any{"type": "object", "properties": map[string]any{}}
		}
		out[i] = map[string]any{
			"name":         t.Function.Name,
			"description":  t.Function.Description,
			"input_schema": params,
		}
	}
	return out
}

func (p *AnthropicProvider) Chat(ctx context.Context, messages []Message, opts ChatOptions) (LLMResponse, error) {
	opts = defaultChatOptions(opts)

	system, converted := toAnthropicMessages(messages)

	payload := map[string]any{
		"model":       p.Model,
		"messages":    converted,
		"temperature": opts.Temperature,
		"max_tokens":  opts.MaxTokens,
	}
	if system != "" {
		payload["system"] = system
	}
	if len(opts.Tools) > 0 {
		payload["tools"] = toAnthropicTools(opts.Tools)
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return LLMResponse{}, err
	}

	req, err := http.NewRequestWithContext(ctx, "POST",
		"https://api.anthropic.com/v1/messages", bytes.NewReader(body))
	if err != nil {
		return LLMResponse{}, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", p.APIKey)
	req.Header.Set("anthropic-version", "2023-06-01")

	t0 := time.Now()
	httpResp, err := p.client.Do(req)
	latencyMs := int(time.Since(t0).Milliseconds())
	if err != nil {
		return LLMResponse{}, err
	}
	defer httpResp.Body.Close()

	respBody, err := io.ReadAll(httpResp.Body)
	if err != nil {
		return LLMResponse{}, err
	}
	if httpResp.StatusCode >= 400 {
		return LLMResponse{}, fmt.Errorf("Anthropic API error %d: %s", httpResp.StatusCode, respBody)
	}

	var parsed struct {
		Content []struct {
			Type  string         `json:"type"`
			Text  string         `json:"text,omitempty"`
			ID    string         `json:"id,omitempty"`
			Name  string         `json:"name,omitempty"`
			Input map[string]any `json:"input,omitempty"`
		} `json:"content"`
		StopReason string `json:"stop_reason"`
		Usage      struct {
			InputTokens  int `json:"input_tokens"`
			OutputTokens int `json:"output_tokens"`
		} `json:"usage"`
	}
	if err := json.Unmarshal(respBody, &parsed); err != nil {
		return LLMResponse{}, fmt.Errorf("parse Anthropic response: %w", err)
	}

	var contentParts []string
	var toolCalls []ToolCall
	for _, block := range parsed.Content {
		switch block.Type {
		case "text":
			contentParts = append(contentParts, block.Text)
		case "tool_use":
			argBytes, _ := json.Marshal(block.Input)
			toolCalls = append(toolCalls, ToolCall{
				ID:       block.ID,
				Type:     "function",
				Function: ToolCallFunction{Name: block.Name, Arguments: string(argBytes)},
			})
		}
	}

	return LLMResponse{
		Content:      strings.Join(contentParts, "\n"),
		ToolCalls:    toolCalls,
		TokenUsage:   TokenUsage{PromptTokens: parsed.Usage.InputTokens, CompletionTokens: parsed.Usage.OutputTokens, TotalTokens: parsed.Usage.InputTokens + parsed.Usage.OutputTokens},
		Model:        p.Model,
		FinishReason: parsed.StopReason,
		LatencyMs:    latencyMs,
	}, nil
}

func (p *AnthropicProvider) SupportsFunctionCalling() bool  { return true }
func (p *AnthropicProvider) SupportsStructuredOutput() bool { return false }
func (p *AnthropicProvider) GetContextWindow() int          { return 200_000 }
func (p *AnthropicProvider) GetModelName() string           { return p.Model }
func (p *AnthropicProvider) CountTokens(text string) int    { return EstimateTokens(text) }

// ---------------------------------------------------------------------------
// Ollama provider  (OpenAI-compatible local endpoint)
// ---------------------------------------------------------------------------

var ollamaToolModels = map[string]bool{
	"llama3.1": true, "llama3.1:8b": true, "llama3.1:70b": true,
	"llama3.2": true, "llama3.2:3b": true,
	"mistral": true, "mistral-nemo": true,
	"qwen2.5": true, "qwen2.5:7b": true,
}

// OllamaProvider uses the OpenAI-compatible endpoint that Ollama exposes.
type OllamaProvider struct {
	inner *OpenAIProvider
	model string
}

// NewOllamaProvider creates an OllamaProvider for the given model.
func NewOllamaProvider(model, baseURL string) *OllamaProvider {
	if baseURL == "" {
		baseURL = "http://localhost:11434/v1"
	}
	inner := NewOpenAIProvider("ollama", model)
	inner.BaseURL = baseURL
	return &OllamaProvider{inner: inner, model: model}
}

func (p *OllamaProvider) Chat(ctx context.Context, messages []Message, opts ChatOptions) (LLMResponse, error) {
	if !p.SupportsFunctionCalling() {
		opts.Tools = nil
	}
	return p.inner.Chat(ctx, messages, opts)
}

func (p *OllamaProvider) SupportsFunctionCalling() bool {
	base := strings.Split(p.model, ":")[0]
	return ollamaToolModels[base] || ollamaToolModels[p.model]
}
func (p *OllamaProvider) SupportsStructuredOutput() bool { return false }
func (p *OllamaProvider) GetContextWindow() int          { return 128_000 }
func (p *OllamaProvider) GetModelName() string           { return p.model }
func (p *OllamaProvider) CountTokens(text string) int    { return EstimateTokens(text) }

// ---------------------------------------------------------------------------
// FallbackProvider
// ---------------------------------------------------------------------------

// FallbackProvider tries providers in order, falling back on any error.
type FallbackProvider struct {
	Primary   LLMProvider
	Fallbacks []LLMProvider
}

func (p *FallbackProvider) Chat(ctx context.Context, messages []Message, opts ChatOptions) (LLMResponse, error) {
	candidates := append([]LLMProvider{p.Primary}, p.Fallbacks...)
	var lastErr error
	for _, provider := range candidates {
		resp, err := provider.Chat(ctx, messages, opts)
		if err == nil {
			return resp, nil
		}
		lastErr = err
		log.Printf("WARN provider %s failed (%v) — trying next", provider.GetModelName(), err)
	}
	return LLMResponse{}, fmt.Errorf("all providers failed; last error: %w", lastErr)
}

func (p *FallbackProvider) SupportsFunctionCalling() bool { return p.Primary.SupportsFunctionCalling() }
func (p *FallbackProvider) SupportsStructuredOutput() bool {
	return p.Primary.SupportsStructuredOutput()
}
func (p *FallbackProvider) GetContextWindow() int { return p.Primary.GetContextWindow() }
func (p *FallbackProvider) GetModelName() string {
	names := []string{p.Primary.GetModelName()}
	for _, f := range p.Fallbacks {
		names = append(names, f.GetModelName())
	}
	return strings.Join(names, " → ")
}
func (p *FallbackProvider) CountTokens(text string) int { return p.Primary.CountTokens(text) }

// ---------------------------------------------------------------------------
// LLMFactory
// ---------------------------------------------------------------------------

// ProviderConfig is a simple map used by LLMFactory.CreateFromConfig.
type ProviderConfig map[string]string

// LLMFactory creates LLMProvider instances from configuration.
type LLMFactory struct{}

// Create builds a provider by name with the given options map.
//
//	opts["api_key"]  — API key (empty for Ollama)
//	opts["model"]    — model identifier
//	opts["base_url"] — override endpoint (optional)
func (LLMFactory) Create(provider string, opts ProviderConfig) (LLMProvider, error) {
	apiKey := opts["api_key"]
	model := opts["model"]

	switch provider {
	case "openai":
		if model == "" {
			model = "gpt-4o"
		}
		p := NewOpenAIProvider(apiKey, model)
		if bu := opts["base_url"]; bu != "" {
			p.BaseURL = bu
		}
		return p, nil
	case "anthropic":
		if model == "" {
			model = "claude-3-5-sonnet-20241022"
		}
		return NewAnthropicProvider(apiKey, model), nil
	case "ollama":
		if model == "" {
			model = "llama3.1:8b"
		}
		return NewOllamaProvider(model, opts["base_url"]), nil
	default:
		return nil, fmt.Errorf("unknown provider %q; available: openai, anthropic, ollama", provider)
	}
}

// CreateFromEnv builds a provider using the named environment variable as API key.
func (f LLMFactory) CreateFromEnv(provider, envKey, model string) (LLMProvider, error) {
	return f.Create(provider, ProviderConfig{
		"api_key": os.Getenv(envKey),
		"model":   model,
	})
}

File: code/nodejs/03-agent-loop/llm_provider.ts
/**
 * Provider-agnostic LLM abstraction layer — TypeScript port.
 *
 * Implements a uniform interface over multiple LLM providers so that agent
 * code never depends on a specific SDK or API format.
 *
 * Supported providers:
 *   OpenAIProvider, AnthropicProvider, GoogleProvider,
 *   OllamaProvider, TogetherProvider, FallbackProvider
 *
 * Usage:
 *   const provider = LLMFactory.create("openai", {
 *     apiKey: process.env.OPENAI_API_KEY!,
 *     model: "gpt-4o",
 *   });
 *   const response = await provider.chat([{ role: "user", content: "Hello" }]);
 *   console.log(response.content);
 *
 * See: docs/05-the-tool-ecosystem/01-model-providers.md
 */

import { performance } from "perf_hooks";

// ---------------------------------------------------------------------------
// Shared data types
// ---------------------------------------------------------------------------

export interface LLMResponse {
  content: string | null;
  toolCalls: ToolCall[] | null;
  tokenUsage: { promptTokens: number; completionTokens: number; totalTokens: number };
  model: string;
  finishReason: string;
  latencyMs: number;
}

export interface ToolCall {
  id: string;
  type: "function";
  function: { name: string; arguments: string };
}

export interface Message {
  role: "system" | "user" | "assistant" | "tool";
  content: string;
  toolCallId?: string;   // for role === "tool"
  toolCalls?: ToolCall[]; // for role === "assistant"
}

export interface ToolDefinition {
  type: "function";
  function: {
    name: string;
    description?: string;
    parameters?: Record<string, unknown>;
  };
}

/** Rough token estimate: ~4 chars per token. */
export function estimateTokens(text: string): number {
  return Math.max(1, Math.floor(text.length / 4));
}

// ---------------------------------------------------------------------------
// Abstract interface
// ---------------------------------------------------------------------------

export abstract class LLMProvider {
  abstract chat(
    messages: Message[],
    options?: {
      tools?: ToolDefinition[];
      temperature?: number;
      maxTokens?: number;
      responseFormat?: Record<string, unknown>;
    }
  ): Promise<LLMResponse>;

  abstract supportsFunctionCalling(): boolean;
  abstract supportsStructuredOutput(): boolean;
  abstract getContextWindow(): number;
  abstract getModelName(): string;

  /** Count tokens. Override with a real tokenizer when available. */
  countTokens(text: string): number {
    return estimateTokens(text);
  }
}

// ---------------------------------------------------------------------------
// OpenAI provider
// ---------------------------------------------------------------------------

const OPENAI_CONTEXT_WINDOWS: Record<string, number> = {
  "gpt-4o":       128_000,
  "gpt-4o-mini":  128_000,
  "gpt-3.5-turbo": 16_385,
};

export class OpenAIProvider extends LLMProvider {
  private client: import("openai").default | null = null;
  private readonly apiKey: string;
  private readonly model: string;
  private readonly baseUrl?: string;

  constructor(opts: { apiKey: string; model?: string; baseUrl?: string }) {
    super();
    this.apiKey = opts.apiKey;
    this.model = opts.model ?? "gpt-4o";
    this.baseUrl = opts.baseUrl;
  }

  private async getClient(): Promise<import("openai").default> {
    if (!this.client) {
      const { default: OpenAI } = await import("openai" as string) as { default: typeof import("openai").default };
      this.client = new OpenAI({
        apiKey: this.apiKey,
        ...(this.baseUrl ? { baseURL: this.baseUrl } : {}),
      });
    }
    return this.client;
  }

  async chat(
    messages: Message[],
    options: {
      tools?: ToolDefinition[];
      temperature?: number;
      maxTokens?: number;
      responseFormat?: Record<string, unknown>;
    } = {}
  ): Promise<LLMResponse> {
    const openai = await this.getClient();
    const t0 = performance.now();

    const params: Record<string, unknown> = {
      model: this.model,
      messages: messages as unknown[],
      temperature: options.temperature ?? 0.7,
      max_tokens: options.maxTokens ?? 4096,
    };
    if (options.tools?.length) params["tools"] = options.tools;
    if (options.responseFormat) params["response_format"] = options.responseFormat;

    const resp = await (openai.chat.completions.create as (p: Record<string, unknown>) => Promise<import("openai").ChatCompletion>)(params);
    const latencyMs = Math.round(performance.now() - t0);
    const msg = resp.choices[0].message;

    const toolCalls: ToolCall[] | null = msg.tool_calls?.map((tc) => ({
      id: tc.id,
      type: "function" as const,
      function: { name: tc.function.name, arguments: tc.function.arguments },
    })) ?? null;

    return {
      content: msg.content,
      toolCalls,
      tokenUsage: {
        promptTokens: resp.usage?.prompt_tokens ?? 0,
        completionTokens: resp.usage?.completion_tokens ?? 0,
        totalTokens: resp.usage?.total_tokens ?? 0,
      },
      model: this.model,
      finishReason: resp.choices[0].finish_reason,
      latencyMs,
    };
  }

  supportsFunctionCalling(): boolean { return true; }
  supportsStructuredOutput(): boolean { return true; }
  getContextWindow(): number { return OPENAI_CONTEXT_WINDOWS[this.model] ?? 128_000; }
  getModelName(): string { return this.model; }
}

// ---------------------------------------------------------------------------
// Anthropic provider
// ---------------------------------------------------------------------------

type AnthropicMessage = { role: "user" | "assistant"; content: unknown };

export class AnthropicProvider extends LLMProvider {
  private client: import("@anthropic-ai/sdk").default | null = null;
  private readonly apiKey: string;
  private readonly model: string;

  constructor(opts: { apiKey: string; model?: string }) {
    super();
    this.apiKey = opts.apiKey;
    this.model = opts.model ?? "claude-3-5-sonnet-20241022";
  }

  private async getClient(): Promise<import("@anthropic-ai/sdk").default> {
    if (!this.client) {
      const { default: Anthropic } = await import("@anthropic-ai/sdk" as string) as { default: typeof import("@anthropic-ai/sdk").default };
      this.client = new Anthropic({ apiKey: this.apiKey });
    }
    return this.client;
  }

  /** Split system message; convert the rest to Anthropic format. */
  private toAnthropicMessages(messages: Message[]): { system?: string; messages: AnthropicMessage[] } {
    let system: string | undefined;
    const converted: AnthropicMessage[] = [];
    for (const msg of messages) {
      if (msg.role === "system") { system = msg.content; continue; }
      if (msg.role === "tool") {
        converted.push({
          role: "user",
          content: [{ type: "tool_result", tool_use_id: msg.toolCallId ?? "", content: msg.content }],
        });
      } else {
        converted.push({ role: msg.role as "user" | "assistant", content: msg.content });
      }
    }
    return { system, messages: converted };
  }

  /** Convert OpenAI tool definitions to Anthropic format. */
  private toAnthropicTools(tools: ToolDefinition[]): unknown[] {
    return tools.map((t) => ({
      name: t.function.name,
      description: t.function.description ?? "",
      input_schema: t.function.parameters ?? { type: "object", properties: {} },
    }));
  }

  async chat(
    messages: Message[],
    options: {
      tools?: ToolDefinition[];
      temperature?: number;
      maxTokens?: number;
      responseFormat?: Record<string, unknown>;
    } = {}
  ): Promise<LLMResponse> {
    const anthropic = await this.getClient();
    const { system, messages: converted } = this.toAnthropicMessages(messages);

    const params: Record<string, unknown> = {
      model: this.model,
      messages: converted,
      temperature: options.temperature ?? 0.7,
      max_tokens: options.maxTokens ?? 4096,
    };
    if (system) params["system"] = system;
    if (options.tools?.length) params["tools"] = this.toAnthropicTools(options.tools);

    const t0 = performance.now();
    type AnthropicResponse = {
      content: Array<{ type: string; text?: string; id?: string; name?: string; input?: Record<string, unknown> }>;
      stop_reason: string;
      usage: { input_tokens: number; output_tokens: number };
    };
    const resp = await (anthropic.messages.create as (p: Record<string, unknown>) => Promise<AnthropicResponse>)(params);
    const latencyMs = Math.round(performance.now() - t0);

    let content: string | null = null;
    const toolCalls: ToolCall[] = [];
    for (const block of resp.content) {
      if (block.type === "text" && block.text) content = (content ?? "") + block.text;
      if (block.type === "tool_use" && block.id && block.name) {
        toolCalls.push({
          id: block.id,
          type: "function",
          function: { name: block.name, arguments: JSON.stringify(block.input ?? {}) },
        });
      }
    }

    return {
      content,
      toolCalls: toolCalls.length ? toolCalls : null,
      tokenUsage: {
        promptTokens: resp.usage.input_tokens,
        completionTokens: resp.usage.output_tokens,
        totalTokens: resp.usage.input_tokens + resp.usage.output_tokens,
      },
      model: this.model,
      finishReason: resp.stop_reason ?? "stop",
      latencyMs,
    };
  }

  supportsFunctionCalling(): boolean { return true; }
  supportsStructuredOutput(): boolean { return false; }
  getContextWindow(): number { return 200_000; }
  getModelName(): string { return this.model; }
}

// ---------------------------------------------------------------------------
// Ollama provider  (OpenAI-compatible local endpoint)
// ---------------------------------------------------------------------------

const OLLAMA_TOOL_MODELS = new Set([
  "llama3.1", "llama3.1:8b", "llama3.1:70b",
  "llama3.2", "llama3.2:3b",
  "mistral", "mistral-nemo",
  "qwen2.5", "qwen2.5:7b",
  "command-r",
]);

export class OllamaProvider extends LLMProvider {
  private readonly inner: OpenAIProvider;
  private readonly model: string;

  constructor(opts: { model?: string; baseUrl?: string }) {
    super();
    this.model = opts.model ?? "llama3.1:8b";
    this.inner = new OpenAIProvider({
      apiKey: "ollama",
      model: this.model,
      baseUrl: opts.baseUrl ?? "http://localhost:11434/v1",
    });
  }

  async chat(messages: Message[], options: {
    tools?: ToolDefinition[];
    temperature?: number;
    maxTokens?: number;
    responseFormat?: Record<string, unknown>;
  } = {}): Promise<LLMResponse> {
    const effectiveTools = this.supportsFunctionCalling() ? options.tools : undefined;
    return this.inner.chat(messages, { ...options, tools: effectiveTools });
  }

  supportsFunctionCalling(): boolean {
    const base = this.model.split(":")[0];
    return OLLAMA_TOOL_MODELS.has(base) || OLLAMA_TOOL_MODELS.has(this.model);
  }
  supportsStructuredOutput(): boolean { return false; }
  getContextWindow(): number { return 128_000; }
  getModelName(): string { return this.model; }
}

// ---------------------------------------------------------------------------
// Together AI provider  (OpenAI-compatible cloud endpoint)
// ---------------------------------------------------------------------------

export class TogetherProvider extends LLMProvider {
  private readonly inner: OpenAIProvider;
  private readonly model: string;

  constructor(opts: { apiKey: string; model?: string }) {
    super();
    this.model = opts.model ?? "meta-llama/Meta-Llama-3.1-70B-Instruct-Turbo";
    this.inner = new OpenAIProvider({
      apiKey: opts.apiKey,
      model: this.model,
      baseUrl: "https://api.together.xyz/v1",
    });
  }

  async chat(messages: Message[], options = {}): Promise<LLMResponse> {
    return this.inner.chat(messages, options);
  }

  supportsFunctionCalling(): boolean { return false; }
  supportsStructuredOutput(): boolean { return false; }
  getContextWindow(): number { return 128_000; }
  getModelName(): string { return this.model; }
}

// ---------------------------------------------------------------------------
// FallbackProvider
// ---------------------------------------------------------------------------

export class FallbackProvider extends LLMProvider {
  private readonly primary: LLMProvider;
  private readonly fallbacks: LLMProvider[];

  constructor(primary: LLMProvider, fallbacks: LLMProvider[]) {
    super();
    this.primary = primary;
    this.fallbacks = fallbacks;
  }

  async chat(messages: Message[], options = {}): Promise<LLMResponse> {
    const candidates = [this.primary, ...this.fallbacks];
    let lastError: Error | null = null;
    for (const provider of candidates) {
      try {
        return await provider.chat(messages, options);
      } catch (err) {
        lastError = err instanceof Error ? err : new Error(String(err));
        console.warn(
          `[FallbackProvider] ${provider.getModelName()} failed: ${lastError.message} — trying next`
        );
      }
    }
    throw new Error(`All providers failed. Last error: ${lastError?.message}`);
  }

  supportsFunctionCalling(): boolean { return this.primary.supportsFunctionCalling(); }
  supportsStructuredOutput(): boolean { return this.primary.supportsStructuredOutput(); }
  getContextWindow(): number { return this.primary.getContextWindow(); }
  getModelName(): string {
    return [this.primary, ...this.fallbacks].map((p) => p.getModelName()).join(" → ");
  }
}

// ---------------------------------------------------------------------------
// LLMFactory
// ---------------------------------------------------------------------------

export type ProviderConfig =
  | { provider: "openai";    apiKey?: string; apiKeyEnv?: string; model?: string; baseUrl?: string }
  | { provider: "anthropic"; apiKey?: string; apiKeyEnv?: string; model?: string }
  | { provider: "ollama";    model?: string; baseUrl?: string }
  | { provider: "together";  apiKey?: string; apiKeyEnv?: string; model?: string }
  | { provider: string;      [key: string]: unknown };

export type ProviderConfigWithFallback = ProviderConfig & {
  fallback?: ProviderConfigWithFallback;
};

export class LLMFactory {
  static create(config: ProviderConfig): LLMProvider {
    const apiKey = "apiKeyEnv" in config && config.apiKeyEnv
      ? (process.env[config.apiKeyEnv] ?? "")
      : ("apiKey" in config ? (config.apiKey ?? "") : "");

    switch (config.provider) {
      case "openai":
        return new OpenAIProvider({
          apiKey,
          model: (config as { model?: string }).model,
          baseUrl: (config as { baseUrl?: string }).baseUrl,
        });
      case "anthropic":
        return new AnthropicProvider({ apiKey, model: (config as { model?: string }).model });
      case "ollama":
        return new OllamaProvider({
          model: (config as { model?: string }).model,
          baseUrl: (config as { baseUrl?: string }).baseUrl,
        });
      case "together":
        return new TogetherProvider({ apiKey, model: (config as { model?: string }).model });
      default:
        throw new Error(
          `Unknown provider: "${config.provider}". Available: openai, anthropic, ollama, together`
        );
    }
  }

  static createFromConfig(config: ProviderConfigWithFallback): LLMProvider {
    const primary = LLMFactory.create(config);
    if (config.fallback) {
      const fallback = LLMFactory.createFromConfig(config.fallback);
      return new FallbackProvider(primary, [fallback]);
    }
    return primary;
  }
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

async function _demo(): Promise<void> {
  const question = "What is 17 × 23? Answer with just the number.";
  const messages: Message[] = [{ role: "user", content: question }];

  const providers: Array<[string, LLMProvider]> = [];

  const oaiKey = process.env["OPENAI_API_KEY"];
  const antKey = process.env["ANTHROPIC_API_KEY"];

  if (oaiKey) {
    providers.push(["OpenAI gpt-4o-mini", LLMFactory.create({ provider: "openai", apiKey: oaiKey, model: "gpt-4o-mini" })]);
  }
  if (antKey) {
    providers.push(["Anthropic claude-3-haiku", LLMFactory.create({ provider: "anthropic", apiKey: antKey, model: "claude-3-haiku-20240307" })]);
  }
  providers.push(["Ollama llama3.1:8b", LLMFactory.create({ provider: "ollama", model: "llama3.1:8b" })]);

  if (!providers.length) {
    console.log("Set OPENAI_API_KEY or ANTHROPIC_API_KEY to run the demo.");
    return;
  }

  console.log("\n=== Provider Comparison ===");
  console.log(`Question: ${question}\n`);
  console.log("Provider".padEnd(30), "Answer".padEnd(20), "Tokens".padEnd(10), "Latency (ms)");
  console.log("-".repeat(75));
  for (const [name, provider] of providers) {
    try {
      const resp = await provider.chat(messages, { temperature: 0, maxTokens: 20 });
      console.log(
        name.padEnd(30),
        (resp.content ?? "").trim().padEnd(20),
        String(resp.tokenUsage.totalTokens).padEnd(10),
        resp.latencyMs
      );
    } catch (err) {
      console.log(name.padEnd(30), `ERROR: ${err}`);
    }
  }

  // Fallback demo
  console.log("\n=== Fallback Demonstration ===");
  class AlwaysFail extends LLMProvider {
    async chat(): Promise<LLMResponse> { throw new Error("Simulated outage"); }
    supportsFunctionCalling() { return false; }
    supportsStructuredOutput() { return false; }
    getContextWindow() { return 0; }
    getModelName() { return "always-fail"; }
  }
  const real = providers[0][1];
  const fallback = new FallbackProvider(new AlwaysFail(), [real]);
  const resp = await fallback.chat(messages, { temperature: 0, maxTokens: 20 });
  console.log(`Fallback result: ${JSON.stringify(resp.content)} (from ${real.getModelName()})`);
}

// Run demo when executed directly
const isMain = process.argv[1]?.endsWith("llm_provider.ts") ||
               process.argv[1]?.endsWith("llm_provider.js");
if (isMain) {
  _demo().catch(console.error);
}
